smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserVisitor.java

```diff
@@ -0,0 +1,433 @@
+ * functions. This may be overwritten by downstream implementations if the protocol requires
+ * more complex deserialization strategies for those types.
+ *
+ */
+public abstract class DocumentShapeDeserVisitor extends ShapeVisitor.Default<Void> {
+    private GenerationContext context;
+    private String protocolName;
+
+    public DocumentShapeDeserVisitor(GenerationContext context, String protocolName) {
+        this.context = context;
+        this.protocolName = protocolName;
+    }
+
+    /**
+     * Gets the generation context.
+     *
+     * @return The generation context.
+     */
+    protected GenerationContext getContext() {
+        return context;
+    }
+
+    /**
+     * Gets the name of the protocol.
+     *
+     * @return The protocol name.
+     */
+    protected String getProtocolName() {
+        return protocolName;
+    }
+
+    @Override
+    protected Void getDefault(Shape shape) {
+        return null;
+    }
+
+    /**
+     * Writes the code needed to deserialize a collection in the document of a response.
+     *
+     * <p>Implementations of this method are expected to generate a function body that
+     * returns the type generated for the CollectionShape {@code shape} parameter from an input
+     * deserialized by {@code deserializeOutputDocument}.
+     *
+     * <p>The function signature for this body will have two parameters available in scope:
+     * <ul>
+     *   <li>{@code output}: a value representing the CollectionShape shape parameter.</li>
+     *   <li>{@code context}: the SerdeContext.</li>
+     * </ul>
+     *
+     * <p>For example, this function would generate the following:
+     *
+     * <pre>{@code
+     * return (output || []).map((entry: any) =>
+     *   deserializeAws_restJson1_1Parameter(entry, context)
+     * );
+     * }</pre>
+     *
+     * @param context The generation context.
+     * @param shape The collection shape being generated.
+     */
+    protected abstract void deserializeCollection(GenerationContext context, CollectionShape shape);
+
+    /**
+     * Writes the code needed to deserialize a map in the document of a response.
+     *
+     * <p>Implementations of this method are expected to generate a function body that
+     * returns the type generated for the MapShape {@code shape} parameter from an input
+     * deserialized by {@code deserializeOutputDocument}.
+     *
+     * <p>The function signature for this body will have two parameters available in scope:
+     * <ul>
+     *   <li>{@code output}: a value representing the MapShape shape parameter.</li>
+     *   <li>{@code context}: the SerdeContext.</li>
+     * </ul>
+     *
+     * <p>For example, this function would generate the following:
+     *
+     * <pre>{@code
+     * let mapParams: any = {};
+     * Object.keys(output).forEach(key => {
+     *   mapParams[key] = deserializeAws_restJson1_1Field(output[key], context);
```

Comment:
Doesn't this need the [hasOwnProperty](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) check to not send random members from the object?

Reply:
Maps are generated as [`{ [key: string]: %s }`](https://github.com/awslabs/smithy-typescript/blob/master/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/SymbolVisitor.java#L161), where %s is the generated type of the map value. As such, they won't have additional properties and the check shouldn't be necessary.